### PR TITLE
[Implements #66] Add Config Class

### PR DESCRIPTION
**Pull request by cascadianblue**
It was very convoluted but it turns out there is [a way](https://www.freecodecamp.org/news/how-to-implement-runtime-environment-variables-with-create-react-app-docker-and-nginx-7f9d42a91d70/) to inject (a subset of) environment variables into the React app at runtime. This implements that, and makes them accessible to the app through a `Config` class interface. Differently from the server-side Config class, this one has its default values stored in `.env`

Specifically, this is useful for setting the client to use an API at a different origin than the hard-coded defaults. I.e. if you are running a dev server, you can have it point to a remote back end by running `SERVER_HOST=phideidentifier.com SERVER_PORT=80 npm run dev`, and this will launch a web client that will rely on `phideidentifier.com` to perform the back end deidentification work.

### CODE REVIEW

File: client/Dockerfile [hunk 1]
```diff
@@ -13,5 +13,14 @@ RUN npm run build
 FROM nginx:stable-alpine
```
Comment (tschaffter) [inline client/Dockerfile:13]: For the sake of reproducibility, we prefer using specific version. Dependabot is configured to let us know when a new version is available.

File: client/package.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
     "web-vitals": "^0.2.4"
   },
   "scripts": {
+    "dev": "chmod +x ./env.sh && ./env.sh && cp env-config.js ./public/ && react-scripts start",
```
Comment (tschaffter) [inline client/package.json:19]: I usually use a verb for the npm scripts. Is this about starting the app in development mode? If yes, then we can have `start:dev`

File: client/package.json [hunk 3]
```diff
@@ -16,6 +16,7 @@
...
     "start": "react-scripts start",
```
Comment (tschaffter) [inline client/package.json:20]: Is this script still needed?

File: client/public/index.html [hunk 4]
```diff
@@ -24,6 +24,7 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
+    <script src="%PUBLIC_URL%/env-config.js"></script>
```
Comment (tschaffter) [inline client/public/index.html:27]: Change the variable name `PUBLIC_URL` to something more descriptive. For example `APP_URL`? Using the keyword PUBLIC does not sound right.

File: client/src/components/App.js [hunk 5]
```diff
@@ -9,8 +9,10 @@ import { encodeString, decodeString } from '../stringSmuggler';
 import { AppBar, Box, IconButton, Toolbar, Typography } from '@material-ui/core';
 import InfoIcon from '@material-ui/icons/Info';
 import { InfoDialog } from './InfoDialog';
+import Config from '../config';
 
-const apiConfiguration = new Configuration({basePath: "http://localhost/api/v1"}); // FIXME: Figure out how to handle hostname
+const config = new Config()
+const apiConfiguration = new Configuration({basePath: config.serverApiUrl()}); // FIXME: Figure out how to handle hostname
```
Comment (tschaffter) [inline client/src/components/App.js:15]: Is this FIXME comment still needed?

File: docker-compose.yml [hunk 6]
```diff
@@ -49,6 +49,8 @@ services:
       context: client
       dockerfile: Dockerfile
     container_name: phi-deidentifier-app
+    environment:
+      - SERVER_PORT=80
```
Comment (tschaffter) [inline docker-compose.yml:53]: We should be able to specify here the SERVER_SCHEME and SERVER_HOST, no?

**Review by tschaffter — CHANGES_REQUESTED** [review]

File: client/public/index.html (hunk 4 above)
Comment (cascadianblue) [reply client/public/index.html:27]: `PUBLIC_URL` is not my variable name, it is the name of a constant [provided by React to provide access to statically-served assets](https://create-react-app.dev/docs/using-the-public-folder/#adding-assets-outside-of-the-module-system)

File: docker-compose.yml (hunk 6 above)
Comment (cascadianblue) [reply docker-compose.yml:53]: Yes, we can, but the defaults in [`.env`](https://github.com/nlpsandbox/phi-deidentifier/blob/6a099bb52f2dae17dfae5970b6daac3215ca24e6/client/.env) are already what we want them to be. Should I include the same values in `docker-compose.yml` anyway?

File: client/package.json (hunk 3 above)
Comment (cascadianblue) [reply client/package.json:20]: No, I don't think it is.

File: client/package.json (hunk 2 above)
Comment (cascadianblue) [reply client/package.json:19]: I just changed it to `start`, actually. Does that work?

File: client/Dockerfile (hunk 1 above)
Comment (cascadianblue) [reply client/Dockerfile:13]: Done! And did same for the Node base image, too.

File: docker-compose.yml (hunk 6 above)
Comment (tschaffter) [reply docker-compose.yml:53]: Yes, let's be more explicit. You can add variables using this synthax: `VARIABLE=${VARIABLE}` where the value `${VARIABLE}` will be read automatically from `.env` by docker-compose.

Comment (cascadianblue) [reply docker-compose.yml]: Something that might not be clear: the file that I had named `client/.env` has nothing to do with the `.env` used to provide values for environmental variables in `docker-compose`. Therefore I renamed the file to `env_vars`.

`client/env_vars` gets read by a script, `client/env.sh`, which then outputs env var values (defaulting to the values in `client/env_vars`) to a static file `env-config.js`, which gets served (statically). Like I said at the top of this PR, it is very weird and convoluted, but it works.

If you want to override the default values in `client/env_vars`, you can do so in the normal way, by putting them in a `.env` file in the project root. I have tested that this works. Does this all make sense?

Comment (tschaffter) [reply docker-compose.yml]: Sort of but not completely. As long as the docker-compose.yml allows us to specify the port, host, etc. we are good.

**Review by tschaffter — APPROVED** [review]

**Pull request merged**